### ISSUE TITLE
Which levels of abstractions in C++ should be exposed to Go

### ISSUE THREAD

**Issue by QiJune** [wangkuiyi/gotorch#57]
There are three levels of abstractions:

- Level 1: native function is a low-level API. There are many basic mathematical operations in it.

- Level 2: nn.functional is a middle-level API. It's more close to deep learning. It uses basic mathematical operations to compose a complex neural network operation.

- Level 3: nn.module is a high-level API. A module contains many states, such as parameters and buffers. It's a C++ class.

Let's take `padding` operator as an example:

- In native function, there is a [`cat` function](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L723)

 - In nn.functional, there is a [`pad` function](https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/include/torch/nn/functional/padding.h#L15), which calls `cat` function in level 1.

- In nn.module, there are [`ZeroPad2d` class](https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/include/torch/nn/modules/padding.h#L212), [`ReplicationPad3d` class](https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/include/torch/nn/modules/padding.h#L175), which call `pad` function in level 2.

|          | expose to Go     | API |  contain state|
|--------------- | --------- |-------|------|
| native function   | C++ function, easy| low-level API, flexible, few users may use it | No |
| nn.functional   | C++ function, easy| middle-level API, most users use it | No |
| nn.module   | C++ class, hard| high-level API, most users use it | Yes, parameters and buffers|

There is another interesting thing, `nn.functional` will try to fuse some basic native functions. Here is an example of [nn.function.linear](https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/include/torch/nn/functional/linear.h#L15).

I am wondering which levels of abstractions in C++ I am supposed to expose to Go?

**Comment by bddppq** [comment]
Here are my thoughts:

- native function: yep and it's a must, because all operators are defined at this level and some of them not exposed in the next two levels. e.g. non "nn" ops, custom ops
- nn functional: nice to have and we should do it, the binding can save us a lot of works of reimplementing the logic with native functions in go.
- nn.module: nice to have, but need to make sure they interact well with go level `torch.Module`. It's probably not easy. e.g. In a model with mixed c++ module (exposed through the binding) and go `torch.Module`, how do we serialize them? if a c++ module takes another module as input (e.g. Sequential), how can we pass a go `torch.Module` to c++?

**Comment by wangkuiyi** [comment]
Thanks for all the kindly reminders and suggestions! @bddppq 

BTW, is it that

1. nn.module and nn.functional are all based on nn.function?
1. is https://github.com/pytorch/xla an alternative implementation of nn.function?

If so, is that we can have gotorch.nn and gotorch.nn.functional just like PyTorch, as long as they all call native functions, and in the future, we can simply replace the implementation of native functions to use https://github.com/pytorch/xla?

**Comment by wangkuiyi** [comment]
We are now clear that we need to migrate three layers from PyTroch to GoTorch:

1. `native functions`: https://github.com/wangkuiyi/gotorch/blob/develop/doc/wrap_native_functions.md
2. Functionals in `torch.nn.functional` and Modules in `torch.nn` : https://github.com/wangkuiyi/gotorch/blob/develop/doc/develop_functionals_and_modules.md
